### PR TITLE
Quickstart focus point change to quick link ribbon

### LINKED ISSUE #16227: Quickstart focus point change to quick link ribbon
Parent #16005 

Implement the Quick Start focus point for Edit your home page, Review Site Pages, and Check your stats in the Quick Link Ribbon to match the Quick Link card.

| Check your site stats | Edit your home page or  Review Site Pages | 
|---|---|
| Screenshot_20220401_165048 | Screenshot_20220401_165052 |

### PR DESCRIPTION
**Pull request by AjeshRPai**
Fixes #16227 

This PR adds the functionality of the quick start focus point on the quick link ribbon if the quick-start card is visible on the Dashboard menu. 

| Quick start focus point in Quick Actions Card  | Quick start focus point in Quick Link Ribbon | 
|---|---|
| Screenshot_20220407_185934 | Screenshot_20220407_190009| 

### Current Implementation 

Please find below the video capture which shows the quick start focus point shown in different scenarios. Due to the limit of mp4 size in GitHub readme, the preview is not shown in the description. Please visit the link 

1. When Tabs feature Disabled - [Video Capture](https://user-images.githubusercontent.com/17463767/162193078-bd9a4b50-392c-4a0a-8078-26142193bf13.mp4)
- All Quickstart focus points should be shown on the Quick Link action card

2. When the Tabs feature is enabled and the default tab is **Site Menu** - [Video Capture](https://user-images.githubusercontent.com/17463767/162192206-9373b7c5-4869-48fc-ae6e-73e754aea0e0.mp4)

- All Quickstart focus points should be shown in the Site Menu tab
- Quick start focus points for Edit your home Page, Review Pages and Check stats should be shown on Quick Link Action Card in the Site Menu Tab  

3.  When the Tabs feature is enabled and the default tab is Dashboard - [Video Capture](https://user-images.githubusercontent.com/17463767/162191361-1b29b136-9ea9-49b2-971e-99c7d042fc7d.mp4)
- Quick start focus points for Edit your home Page, Review Pages and Check stats should be shown on Quick Link Ribbon in the Dashboard Tab  
- Quick start focus points for Check your Site, Social Sharing and Explore plans should show an intermediary step in site menu and should show the site menu item 

### To test:

Test 1 
- Launch the app with a site having a quick start in progress 
- Go to My Site -> Me -> App Settings -> Debug settings.
- Turn off the MySiteDashboardTabsFeatureConfig flag under the Features in development section.
Restart the app.
- Logout/login and select "Show me around" when the quick start prompt is shown.
- Run tests in [wpandroid-quick-start-testing-instructions.txt](https://github.com/wordpress-mobile/WordPress-Android/files/8246096/wpandroid-quick-start-testing-instructions.txt).
- Notice that for the following tasks - **Edit your home Page**, **Review Pages** and **Check stats**, the QuickStart focus point is shown in Quick links 

Test 2 & 3 - Prerequisites
- Go to My Site -> Me -> App Settings -> Debug settings.
- Enable `MySiteDashboardTabsFeatureConfig`
- Restart the app

   
Test 2 - When the default tab is `Site Menu` 
- Navigate to `My Site` -> `Me` -> `App Settings` 
- Tap `Initial Screen` and select `Site Menu` 
- Launch the app with a site having a quick start in progress 
- Run tests in [wpandroid-quick-start-testing-instructions.txt](https://github.com/wordpress-mobile/WordPress-Android/files/8246096/wpandroid-quick-start-testing-instructions.txt).
- Notice that for the following tasks - **Edit your home Page**, **Review Pages** and **Check stats**, the QuickStart focus point is shown in the Quick links Actions Card. 
- Notice that for the quick start tasks  - **Check your Site, Social Sharing and Explore plans**, the Quickstart focus points are shown in the Site menu items as same as before. 

Test 3 - When the default tab is `Home`
- Navigate to `My Site` -> `Me` -> `App Settings` 
- Tap `Initial Screen` and select `Home` 
- Launch the app with a site having a quick start in progress 
- Run tests in [wpandroid-quick-start-testing-instructions.txt](https://github.com/wordpress-mobile/WordPress-Android/files/8246096/wpandroid-quick-start-testing-instructions.txt).
- Notice that for the following tasks - **Edit your home Page**, **Review Pages** and **Check stats**, the QuickStart focus point is shown in the Quick links Ribbon Card. 
- Notice that for the quick start tasks  - **Check your Site, Social Sharing and Explore plans**, the Quickstart focus points is shown in the Site menu tab header first
- On clicking the Site menu tab header, the list gets scrolled to show the Site menu item corresponding to the task 

## Regression Notes
1. Potential unintended areas of impact
Quick start focus points not being shown properly 
Quick start not working properly 

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing and Unit testing

3. What automated tests I added (or what prevented me from doing so)
Updated tests in MySiteViewModelTest

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt [hunk 1]
```diff
@@ -349,15 +349,31 @@ class MySiteViewModel @Inject constructor(
...
 
-    private fun getPositionOfQuickStartItem(siteItems: Map<MySiteTabType, List<MySiteCardAndItem>>) =
+    private fun getPositionOfQuickStartItem(
+        siteItems: Map<MySiteTabType, List<MySiteCardAndItem>>,
+        activeTask: QuickStartTask
+    ) =
             if (isMySiteTabsEnabled) {
-                (siteItems[MySiteTabType.SITE_MENU] as List<MySiteCardAndItem>)
-                        .indexOfFirst { it.activeQuickStartItem }
+                val currentTab = orderedTabTypes[_selectTab.value!!.peekContent().position]
```
Comment (zwarm) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: 👋  @AjeshRPai - It's preferable not to use double bang if there are other approaches to managing nullability. Can this be reworked to use something like the following?

```
_selectTab.value?.let {
  ...
}
```

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt [hunk 2]
```diff
@@ -349,15 +349,31 @@ class MySiteViewModel @Inject constructor(
         }
     }
 
-    private fun getPositionOfQuickStartItem(siteItems: Map<MySiteTabType, List<MySiteCardAndItem>>) =
+    private fun getPositionOfQuickStartItem(
+        siteItems: Map<MySiteTabType, List<MySiteCardAndItem>>,
+        activeTask: QuickStartTask
+    ) =
             if (isMySiteTabsEnabled) {
```
Comment (zwarm) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: Nit Pick: The `if` can move up to the previous line and the fun can be reformatted, that will decrease the indent of the `.indexOfFirst` below

**Review by zwarm — CHANGES_REQUESTED** [review]
@AjeshRPai - Very nice work. Focus point works as expected on both tabs. I left one change request regarding !! and one nit pick. Thanks for all the hard work, I know this was a tricky one.

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 1 above)
Comment (AjeshRPai) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: Yup right. Missed this. I have made the changes in the [09e34c3](https://github.com/wordpress-mobile/WordPress-Android/pull/16267/commits/09e34c3972e4279cb6dbb50cfe94dda28bee0a1e). Please have a look.

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 2 above)
Comment (AjeshRPai) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: @zwarm Yeah, good point.  I have moved up the `if` and formatted the funnction. Please find the changes in 09e34c3. 👍🏼

**Review by zwarm — APPROVED** [review]
@AjeshRPai - All works as expected. Thank you ❤️  for making those changes!!

**Comment by tiagomar** [comment]
👋  @AjeshRPai @zwarm - The happy paths works as expected. However, in a few tasks, if the user change tabs (it could be either intentionally or accidentally) we are not indicating that the current focus point is in the other tab as we do for some of the tasks like `Customize your site > Visit your site`.

The issue will only happen when a tab different from the Initial Screen is selected before the user clicking `GO` to start the affected tasks. The issue is not reproducible when selecting a specific task from the list, since Quick Start card will only appear in the selected Initial Screen.

Impacted tasks identified so far: 
* `Customize your site > Edit your homepage`
* `Customize your site > Review site pages`
* `Grow your audience > Check your stats`

Expected behavior (reference task): `Customize your site > Visit your site`

**Steps**

Initial Screen: `Site Menu`
1. Complete `Customize your site > Check your site title`.
2. Tap `GO` to start the next task.
3. Complete `Customize your site > Choose a unique site icon`.
4. Tap `HOME` tab.
5. Tap `GO` to start the next task.
6. ❌  Notice that no focus point is displayed.
7. Tap `SITE MENU` tab.
8. Notice that the focus point is displayed over `Pages` quick link.


Initial Screen: `HOME`
1. Complete `Customize your site > Check your site title`.
2. Tap `GO` to start the next task.
3. Complete `Customize your site > Choose a unique site icon`.
4. Tap `SITE MENU` tab.
5. Tap `GO` to start the next task.
6. ❌  Notice that no focus point is displayed.
7. Tap `HOME` tab.
8. Notice that the focus point is displayed over `Pages` quick link ribbon card.

**Comment by tiagomar** [comment]
I've filed #16288 regarding the issue in my previous comment.

**Pull request merged**